### PR TITLE
raftstore: optimize the batching mechanism when enabling async-ios.

### PR DESCRIPTION
**Pull request by LykxSassinator**
### What is changed and how it works?

Issue Number: Ref https://github.com/tikv/tikv/issues/16907

What's Changed:

Previously, #https://github.com/tikv/tikv/pull/16906 had already lowered the compression threshold to alleviate the strain on IO resources. However, it was found that this configuration adjustment had limited impact on reducing IO resource costs.

Therefore, this PR introduces a simple algorithm (using sliding window) to track the real-time size of WriteBatch and initiate a `spin` operation if the WriteBatch size is under the desired threshold. This mechanism aims to increase the WriteBatch size as needed, a strategy that has proven effective in saving IO resources.

```commit-message
Introduce a simple algorithm to make the batching mechanism on WriteBatch larger as expected,
for reducing the cost on IO resources.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

By testing on `oltp-read-write` workload:
- It can greatly reduce the cost of IOPS, at most 30%. And relative bandwidth costs can be reduced from 260 MB/s to ~170 MB/s, which will have significant effects on reducing the IO resources cost on Cloud env.
- The performance degression is low as expected < 1%, introduced by the new batching mechanism. 

image

Side effects

- [x] Performance regression: Consumes more CPU
The performance degression is low as expected < 1%, introduced by the new batching mechanism. 
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
Introduce a simple algorithm to make the batching mechanism on WriteBatch larger as expected,
for saving the cost on IO resources.
```

### CODE REVIEW

**Comment by glorv** [comment]
An intuitive idea: What about move write task batch to raft fsm side. Instead of send a write message after handle each region, we can first cache region write message at the PollCtx, and after handle several regions, when the total write msg size exceeds a certain threshold or the current poll round is over, then we send the cached write tasks to the write thread. It seems more easier to handle this batch logic at the raft fsm side than the write io thread side.
/cc @Connor1996 @overvenus What do you think?

File: components/raftstore/src/store/async_io/write.rs [hunk 1]
```diff
@@ -373,6 +375,105 @@ impl<W: WriteBatch, L: RaftLogBatch> ExtraBatchWrite<W, L> {
     }
 }
 
+/// WriteTaskBatchRecorder is a sliding window, used to record the batch size
+/// and calculate the yield interval.
+/// If the batch size is smaller than the threshold, it will return a
+/// recommended yield duration for the caller as a hint to wait for more writes.
+/// The yield interval is calculated based on the trend of the change of the
+/// batch size. The range of the trend is [0.5, 2.0]. If the batch size is
```
Comment (zhangjinpeng87) [inline components/raftstore/src/store/async_io/write.rs:383]: Can you explain what does 2.0 mean and what does 0.5 mean? What is the quantitive relationship with wait time?

Comment (LykxSassinator) [reply components/raftstore/src/store/async_io/write.rs:383]: Just a heuristic ratio according to testing records, reflecting that the time of `yield` ranges from [`yield_time` * 0.5,  `yield_time` * 2.0] nanoseconds. By default, the yield_time is `50 nanoseconds`. So, the ranges of the whole yielding is `[25 nanoseconds, 100 nanoseconds]`.

**Comment by Connor1996** [comment]
> An intuitive idea: What about move write task batch to raft fsm side. Instead of send a write message after handle each region, we can first cache region write message at the PollCtx, and after handle several regions, when the total write msg size exceeds a certain threshold or the current poll round is over, then we send the cached write tasks to the write thread. It seems more easier to handle this batch logic at the raft fsm side than the write io thread side. /cc @Connor1996 @overvenus @zhangjinpeng87 What do you think?

Looks promising, it's just how sync does that batch the writes in one poll. Maybe you can make a quick poc.

File: components/raftstore/src/store/config.rs [hunk 2]
```diff
@@ -327,6 +327,18 @@ pub struct Config {
...
+    /// will yield for a while to make the writebatch larger, which will reduce
+    /// the write amplification.
+    #[doc(hidden)]
+    pub raft_write_batch_size_hint: ReadableSize,
+
+    /// When the size of raft db writebatch is smaller than this value, write
+    /// will yield for a while. This is used to reduce the write amplification.
+    /// Unit: nanoseconds.
+    #[doc(hidden)]
+    pub raft_write_yield_interval: u64,
```
Comment (zhangjinpeng87) [inline components/raftstore/src/store/config.rs:340]: `raft_write_wait_dur` is a better name, users don't understand what does yield mean. Please also update the comments:
"When the size of raft write batch is smaller than batch-size-hint value, write will wait for a while for a larger batch ..."

**Comment by zhangjinpeng87** [comment]
> An intuitive idea: What about move write task batch to raft fsm side. Instead of send a write message after handle each region, we can first cache region write message at the PollCtx, and after handle several regions, when the total write msg size exceeds a certain threshold or the current poll round is over, then we send the cached write tasks to the write thread. It seems more easier to handle this batch logic at the raft fsm side than the write io thread side. /cc @Connor1996 @overvenus @zhangjinpeng87 What do you think?

@LykxSassinator @glorv This is a good idea. I think we can merge current solution first because it is well tested and has promising benefits of reducing IOPS and IO bandwidth, and then evaluate the new idea. What do you think?

File: components/tikv_util/src/time.rs [hunk 3]
```diff
@@ -532,6 +532,13 @@ impl Default for ThreadReadId {
     }
 }
 
+pub fn yield_at_least(elaspsed: Duration) {
```
Comment (glorv) [inline components/tikv_util/src/time.rs:535]: Please add a comment for this function, using `yield` to wait is only fit for very short period, so caller should be cautious when uses this function.

**Review by glorv — COMMENTED** [review]
rest LGTM

**Comment by LykxSassinator** [comment]
Based on the most recent performance comparisons with version 8.1.0 (default LTS version), it appears that the larger batch flushing introduced by this PR may lead to performance regressions in the `oltp_update_non_index` workload.

Therefore, this PR will be put on hold until a more suitable configuration for optimizing the WriteBatch size is identified.

**Comment by LykxSassinator** [comment]
Thx for the recommendations from @glorv and @zhangjinpeng87 , after replacing `yield_now` with `spin` (mocked by looping finite count), the regression has been greatly inhibit on `oltp_update_non_index` workload and the feedback on reducing the cost of IOPS keeps the same as expected.

image
image

**Review by glorv — COMMENTED** [review]
LGTM

File: components/tikv_util/src/time.rs [hunk 4]
```diff
@@ -532,6 +535,63 @@ impl Default for ThreadReadId {
     }
 }
 
+/// Yield the CPU to other threads or processes for at least `elaspsed`
+/// duration synchronously.
+///
+/// Attention, this function is only suitable for short-time spinning, so
+/// the `elaspsed` should be small, like 1ms. And the caller should not
+/// rely on it to guarantee the exact time to sleep.
+pub fn yield_at_least(elaspsed: Duration) {
```
[the remark this answers is not in the document]
Comment (LykxSassinator) [reply components/tikv_util/src/time.rs:544]: Just keep it for further development.  It‘s ok to remove it to keep the code clean.

**Review by SpadeA-Tang — APPROVED** [review]

**Review by glorv — APPROVED** [review]

**Comment by v01dstar** [comment]
Recommend check the performance regression box in the "Side Effect" section of the PR's description.

**Pull request merged**